controllers/space/space_controller.go

```diff
@@ -0,0 +1,345 @@
+	// look-up the NSTemplateTier
+	tmplTier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: space.Namespace,
+		Name:      space.Spec.TierName,
+	}, tmplTier); err != nil {
+		return nil, err
+	}
+	// create the NSTemplateSet from the NSTemplateTier
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: memberOperatorNS,
+			Name:      space.Name,
+			Finalizers: []string{
+				toolchainv1alpha1.FinalizerName,
+			},
+		},
+		Spec: toolchainv1alpha1.NSTemplateSetSpec{
+			TierName: space.Spec.TierName,
+			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{
+				TemplateRef: tmplTier.Spec.ClusterResources.TemplateRef,
+			},
+			Namespaces: make([]toolchainv1alpha1.NSTemplateSetNamespace, len(tmplTier.Spec.Namespaces)),
+		},
+	}
+	for i, tmpl := range tmplTier.Spec.Namespaces {
+		nsTmplSet.Spec.Namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
+			TemplateRef: tmpl.TemplateRef,
+		}
+	}
+
+	return nsTmplSet, nil
+}
+
+func (r *Reconciler) ensureSpaceDeletion(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	logger.Info("terminating Space")
+	if deleted, err := r.deleteNSTemplateSet(logger, space); err != nil {
+		logger.Error(err, "failed to delete the NSTemplateSet")
+		return false, r.setStatusTerminatingFailed(space, "failed to delete the NSTemplateSet")
+	} else if deleted {
+		if err := r.setStatusTerminating(space); err != nil {
+			logger.Error(err, "error updating status")
+			return false, err
+		}
+		return true, nil // requeue until NSTemplateSet is fully deleted
+	}
+	// Remove finalizer from Space
+	util.RemoveFinalizer(space, toolchainv1alpha1.FinalizerName)
+	if err := r.Client.Update(context.TODO(), space); err != nil {
+		logger.Error(err, "failed to remove finalizer")
+		return false, r.setStatusTerminatingFailed(space, "failed to remove finalizer")
+	}
+	logger.Info("removed finalizer")
+	// no need to update the status of the Space once the finalizer has been removed, since
+	// the resource will be deleted
+	return false, nil // no need to requeue
+}
+
+// deleteNSTemplateSet triggers the deletion of the NSTemplateSet on the target member cluster.
+// Returns `true/nil` if the NSTemplateSet was deleted during this call, or if it is already being deleted
+// (ie, the deletion was triggered during a previous call/reconcile loop, and it's not fully done yet)
+// Returns `false/nil` if the NSTemplateSet was already deleted (ie, resource doesn't exist anymore),
+//   or if there is no target cluster specified in the given space, of if the target cluster is unknown.
+// Returns `false/error` if an error occurred
+func (r *Reconciler) deleteNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	if space.Spec.TargetCluster == "" {
```

Comment:
Isn't the targetcluster field optional? If it's empty does it mean the NSTemplateSet could still be in one of the member clusters?

Reply:
no really. The `targetCluster` is an optional field so the UI (or who will create Spaces for AppStudio workspaces) doesn't have to care about capacity of member clusters. There is a plan to create a `SpaceCompletion` controller that will check the capacity of member clusters and will assign the `targetCluster` to those `Spaces` that don't have it set yet: https://issues.redhat.com/browse/CRT-1333
This means that the NSTemplateSet won't be created until the `targetCluster` is set.

In DevSandbox case, the `targetCluster` will be always set by `UserSignup` controller, so we don't have to deal with it right now. This is only for the case of AppStudio so UI can just simply create `Space` without specifying `targetCluster` and let the controllers deal with it.

Reply:
Cool, thanks for the detailed explanation!